closed/src/java.base/share/classes/sun/security/rsa/NativeRSACore.java

```diff
@@ -138,17 +145,57 @@ public static int getByteLength(BigInteger b) {
             outputLen = getByteLength(n);
             output = new byte[outputLen];
             int verifyInt = verify ? outputLen : -1;
-            outputLen = nativeCrypto.RSADP(msg, msg.length, output, verifyInt, nativePtr);
+            SoftReference<RSADPBigNumRPointer> rsadpBnSoftPtr = rsadpBnPtr.get();
+            RSADPBigNumRPointer rsadpBn = rsadpBnSoftPtr == null ? null : rsadpBnSoftPtr.get();
+            if (rsadpBn == null) {
+                try {
+                    rsadpBn = new RSADPBigNumRPointer();
+                }
+                catch(OutOfMemoryError e) {
+                    return null;
+                }
+                rsadpBnPtr.set(new SoftReference<>(rsadpBn));
+            }
+            outputLen = nativeCrypto.RSADP(msg, msg.length, output, verifyInt, nativePtr, rsadpBn.getBN());
+            Reference.reachabilityFence(rsadpBn);
         } finally {
             key.returnNativePtr(nativePtr);
         }
-
         if (outputLen == -1) {
-            return null;
-        } else if (outputLen == -2) {
             throw new BadPaddingException("RSA private key operation failed");
         }
 
         return output;
     }
+
+    static class RSADPBigNumRPointer {
+
+        long rsadpBN = 0;
+        private static final Cleaner rsadpPtrCleaner = CleanerFactory.cleaner();
+
+        RSADPBigNumRPointer() throws OutOfMemoryError {
```

Comment:
Where are the strong references to instances of this class? I see none, in which case they may be collected prematurely leaving a dangling reference to the native memory that was release by the cleaner thread.

Reply:
@keithc-ca Yes, that is correct. When we reach the `reachabilityFence`, we already have all the needed data and the native `RSADPBigNumRPointer` can be safely reclaimed.

Sure, I changed the code to call `create_rsadp_BN()` before constructing the `RSADPBigNumRPointer`.